### ISSUE TITLE
Implement CalculateRollupFieldRequest

### ISSUE THREAD

**Issue by ThabisoMaja** [jordimontana82/fake-xrm-easy#56]
is CalculateRollupFieldRequest supported?

**Comment by jordimontana82** [comment]
HI @ThabisoMaja ,

Not yet. The current list of fake message executors is available [here] (https://github.com/jordimontana82/fake-xrm-easy/tree/master/FakeXrmEasy.Shared/FakeMessageExecutors)

Would you like to add it there?

**Comment by ThabisoMaja** [comment]
Hi @jordimontana82,

Thanks for the feedback. 
How easy is it to add?

**Comment by jordimontana82** [comment]
Based on [this](https://www.youtube.com/watch?v=RoahCH1p3T8&feature=youtu.be) video, the aggregations are basically Sum, Min, and Max. So, it shouldn't be much complicated to have a message that executes those aggregations against the list of entities using [Linq](http://stackoverflow.com/questions/7105505/linq-aggregate-algorithm-explained).

That could be one way of doing it...

**Comment by jordimontana82** [comment]
You also have .Sum(), .Max(), .Min() and so on....

**Comment by jordimontana82** [comment]
Another option would be to use .AddExecutionMock() with pretty much an empty method there, if you are in a hurry, just so that it won't raise an exception and you can test other code. However, the implementation with Linq aggregations would be ideal. :)

What do you think?

**Comment by ThabisoMaja** [comment]
@jordimontana82 thanks for the info, in a bit of a hurry but when i have sometime will at implementing the Linq aggregation.

**Comment by jordimontana82** [comment]
Ok @ThabisoMaja , you're welcome.

**Comment by jordimontana82** [comment]
Hi @ThabisoMaja ,

Just wondering if you had any chance to look at this? :)

**Comment by ThabisoMaja** [comment]
Hi @jordimontana82 

Not yet, bit hectic at the moment but will let you once i have a chance. Thanks

**Comment by mike9182** [comment]
This is something I would like to see built in as well. Although, I am wondering how the Fake Context would determine what calculation steps should be used for the rollup.

**Comment by jordimontana82** [comment]
Hi @mike9182 ,

That message expects just the Target and the field name. So we'll need to retrieve the rollup field definition from the CRM metadata, because in the ProxyTypesAssembly we just have types.  

That is actually in the RetrieveEntityRequest fake executor. So we could inject the required rollup field definition via [that property](https://github.com/jordimontana82/fake-xrm-easy/blob/a43cb03fdd1a3e2d7d8a519bd835f5f1ab5133a9/FakeXrmEasy.Shared/FakeMessageExecutors/RetrieveEntityRequestExecutor.cs#L14-L15).

Then, the CalculateRollupFieldRequest could use the rollup field definition ([obtained from the SourceTypeMask](https://msdn.microsoft.com/en-us/library/dn817863.aspx#BKMK_SourceTypeMasks)) enumeration to carry out the required calculations.

So to recap, one possible solution:

1) Add a new fake message executor to cater for the CalcullateRollupFieldRequest
2) The executor would do RetrieveEntityRequest internally to retrieve the rollup metadata.
3) Check that the attribute is actually a Rollup field (SourceTypeMask)
4) Retrieve the associated XAML FormulaDefinition
5) Implement the calculation based on the Formula Definition

Now... that's plenty of work :)

Would you like to implement it?

**Comment by jordimontana82** [comment]
Or... maybe just a simpler solution would be to just make sure the CalculateFieldRollupRequest was called for a specific Target and field.

**Comment by phuocle** [comment]
Any news update @jordimontana82 because my code now use CalculateRollupFieldRequest and we need test it.

**Comment by DigitalFlow** [comment]
@phuocle:
If you need it right now, consider using AddFakeMessageExecutor for dynamically adding an implementation.

**Comment by jordimontana82** [comment]
If you could, the easiest thing would be to add an executor that just stores your request so that at least, you could check it was called for a specific field name. Or just use directly FakeItEasy in the meantime, but if we store the original request might be easier to query than with FakeItEasy syntax.

**Comment by phuocle** [comment]
Hi @DigitalFlow, after I add ```CalculateRollupFieldRequestExecutor``` to ```fakedContext```, I get an exception, please help me what I can doing next ?
```
An exception of type 'System.MissingMethodException' occurred in FakeXrmEasy.dll but was not handled in user code

Additional information: Method not found: 'FakeItEasy.Configuration.IAfterCallSpecifiedWithOutAndRefParametersConfiguration FakeItEasy.ReturnValueConfigurationExtensions.ReturnsLazily
```

My ```CalculateRollupFieldRequestExecutor``` class
```C#
public class CalculateRollupFieldRequestExecutor : IFakeMessageExecutor
    {
        public bool CanExecute(OrganizationRequest request)
        {
            return request is CalculateRollupFieldRequest;
        }

        public Type GetResponsibleRequestType()
        {
            return typeof(CalculateRollupFieldRequest);
        }

        public OrganizationResponse Execute(OrganizationRequest request, XrmFakedContext ctx)
        {
            var req = request as CalculateRollupFieldRequest;
            var account = new Entity("account")
            {
                ["accountid"] = Guid.NewGuid(),
                ["pl_actualamount"] = new Microsoft.Xrm.Sdk.Money(123)
            };
            var results = new ParameterCollection {{"Entity", account}};
            var response = new CalculateRollupFieldResponse()
            {
                ResponseName = "CalculateRollupField",
                Results = results
            };
            return response;
        }
    }
```

**Comment by DigitalFlow** [comment]
Hi @phuocle,
Does it work like this?

```C#
public void Test() 
{
var context = new XrmFakedContext();

context.AddExecutionMock<CalculateRollupFieldRequest>(request => 
{
var req = request as CalculateRollupFieldRequest; 
var account = new Entity("account") { 
["accountid"] = Guid.NewGuid(), 
["pl_actualamount"] = new Microsoft.Xrm.Sdk.Money(123) 
}; 
var results = new ParameterCollection {{"Entity", account}}; 
var response = new CalculateRollupFieldResponse() { 
ResponseName = "CalculateRollupField",
Results = results
};
return response;
}
}
```

**Comment by phuocle** [comment]
Hi @DigitalFlow, everything now worked. Thanks for support.

**Comment by mikejramsey** [comment]
The (_inherited_) project I am working on is using a dated version of FakeXrmEasy.2016  (version 1.29.1) and I am running into this issue with the CalculateRollupFieldRequest not implemented. I am curious if this issue has been resolved and implemented in a more current version that I can upgrade to.

**Comment by jordimontana82** [comment]
Hi Mike, it's not there yet, however I think it would be easier to just add an execution mock like @DigitalFlow  pointed out above or else, add your own fake message with 

     context.AddFakeMessageExecutor<Request>(new YourOwnFakeMessageExecutorForThatRequest())

any fake message included in FakeXrmEasy can be overriden using any of the 2 methods.

Hope this helps.

**Comment by jordimontana82** [comment]
Closing this one as the complexity of actually implementing rollup fields is not so trivial and there is the workaround of injecting your own mocks, now documented in [wiki](https://github.com/jordimontana82/fake-xrm-easy/wiki/Mock-an-OrganizationRequest)
